File: 04-stack/00155-Min_Stack/00155-min_stack.cpp
// 155: Min stack
// https://leetcode.com/problems/min-stack/


#include<iostream>
#include<stack>

using namespace std;

// SOLUTION
class MinStack {
    private:
        stack<int> s1, s2;
    
    public:
        void push(int x) {
            s1.push(x);
            if (s2.empty() || x <= getMin())
                s2.push(x);
        }

        void pop() {
            if (s1.top() == getMin())
                s2.pop();
            s1.pop();
        }

        int top() {
            return s1.top();
        }

        int getMin() {
            return s2.top();
        }
};


int main() {
    MinStack *o = new MinStack();

    // OPERATIONS
    o->push(-2);
    o->push(0);
    o->push(-3);
    cout<<o->getMin()<<" ";
    o->pop();
    o->top();
    cout<<o->getMin()<<" ";

    cout<<endl;

    return 0;
}

File: 04-stack/00155-Min_Stack/00155-min_stack.py
# 155: Min stack
# https://leetcode.com/problems/min-stack/


# SOLUTION
class MinStack:
    def __init__(self):
        self.s1 = []
        self.s2 = []
    
    def push(self, x: int)  -> None:
        self.s1.append(x)
        if len(self.s2)==0 or x <= self.getMin():
            self.s2.append(x)
    
    def pop(self) -> None:
        if self.s1[-1] == self.getMin():
            self.s2.pop()
        self.s1.pop()
    
    def top(self) -> int:
        return self.s1[-1]
    
    def getMin(self) -> int:
        return self.s2[-1]



if __name__ == "__main__":
    o = MinStack()
    
    # OPERATIONS
    o.push(-2)
    o.push(0)
    o.push(-3)
    print(o.getMin(), end=" ")
    o.pop()
    o.top()
    print(o.getMin(), end=" ")

    print()